phpseclib/Crypt/Random.php

```diff
@@ -260,6 +260,44 @@ function crypt_random_string($length)
     }
 }
 
+if (!function_exists('phpseclib_safe_serialize')) {
+    /**
+     * Safely serialize variables
+     *
+     * If a class has a private __sleep() method it'll give a fatal error on PHP 5.2 and earlier.
+     * PHP 5.3 will emit a warning.
+     *
+     * @param mixed $arr
+     * @param array $refs optional
+     * @access public
+     */
+    function phpseclib_safe_serialize($arr, $refs = array())
+    {
+        if (is_object($arr)) {
+            return '';
+        }
+        if (!is_array($arr)) {
+            return serialize($arr);
+        }
+        $safearr = array();
+        foreach (array_keys($arr) as $key) {
+            if (is_object($arr[$key])) {
+                continue;
+            }
+            // without this recursive arrays (eg. $a = []; $a[] = &$a;) would result in an
+            // "Allowed memory size of ... bytes exhausted" Fatal error
+            foreach ($refs as $ref) {
+                if ($ref === $arr[$key]) {
```

Comment:
This comparison does not seem to work with older PHP versions. When tested using:

```
[copy-paste of the phpseclib_safe_serialize() definition]
echo phpseclib_safe_serialize($GLOBALS);
```

on http://sandbox.onlinephpfunctions.com/, it outputs following error:

```
<br />
<b>Fatal error</b>:  Nesting level too deep - recursive dependency? in <b>[...][...]</b> on line <b>19</b><br />
```

on the following PHP versions:
5.4.4
5.3.14
5.2.17
5.1.6
5.0.5
4.4.9

Maybe a recursion depth limitation could be used instead? (Like, e.g., in https://secure.php.net/manual/en/function.var-dump.php#76072 or https://secure.php.net/manual/en/function.print-r.php#78851 .)

Note: The behaviour changes in PHP 5.3.15 and 5.4.5, where the error no longer occurs (including later versions). Both were released on 19 Jul 2012, according to https://secure.php.net/releases/.